### ISSUE TITLE
Updated Results Output

### ISSUE THREAD

**Issue by rrojas350** [BrightSpots/rcv#633]
RCTab's result file could use some improvement, specifically a breakdown of inactive ballots. See this mock-up for an example: 

[RCTab Results Report Draft](https://docs.google.com/spreadsheets/d/1YPScnGZnQvJw6klg6CiIXRBCN4v9SEaa_pf7Hsj-ITk/edit?usp=sharing)

**Comment by rrojas350** [comment]
Column C and D have additional information as well. Column C has a percentage of the vote in that round and D has the amount of votes transferred over. Ideally each round would have both the percentage and votes transferred column.

**Comment by chughes297** [comment]
And the creation of the undervote category in B14. This is ballots that have no ranking at all in a contest. Lots of statute coming online now that treats undervotes as a separate category from ballots that have any kind of marking/ranking in an RCV contest (even if it's just an overvote in the first round).

**Comment by ggilber1** [comment]
Also, Column F should be "% of Vote" and G "Transfered"

**Comment by ggilber1** [comment]
Not sure exactly what "Truncated Ballots" are. ?? @rrojas350

**Comment by chughes297** [comment]
@ggilber1 Truncated ballot is a ballot that does not use all of its rankings and goes inactive when it runs out of those rankings. Open to other terms!!

**Comment by ggilber1** [comment]
I would include "truncated" with the "Exhausted Choices" category.  I've
never seen the two separated and don't think the increased complexity is
worth the information gained.

George Gilbert
***@***.***
336-906-0047

**Comment by chughes297** [comment]
Also not sure if this is clear but I'm assuming we'd update the .json file with this sort of info as well, I just didn't make a mockup of that format.

**Comment by artoonie** [comment]
Would this be a good time to include a per-round threshold too, since we're making this update?
I'd like to also add a version number to the output summary files.

**Comment by chughes297** [comment]
> Would this be a good time to include a per-round threshold too, since we're making this update? I'd like to also add a version number to the output summary files.

Yes! That's on row 26. (bottom of candidate totals). Given your question, though, is that easy for a reader to miss? 

Great idea on adding version number too.

**Comment by artoonie** [comment]
Ah, I see it now -- I saw the original threshold on Line 7 and assumed there was no other. The top one on Line 7 is useful for a quick glance, so let's keep both.

I'm making some additional minor adjustments:
1. Since the Inactive Ballots section corresponds with the same headers as the Round-by-Round results, lets group them together so it makes it easier to parse and graph. CSVs don't handle headers well.
2. Undervotes is really just an inactive ballot that started as inactive on Round 1, right? Instead of keeping it separate, why not group it with Inactive Ballots?

I've made these changes and bolded them in this doc: https://docs.google.com/spreadsheets/d/10ZDn-CZK1f98lLZvct3ZqsuJwaLML29tvd3XAM_8nkk/edit#gid=1599108619

**TL;DR can "Undervotes" be a part of "Inactive Ballots" or am I misunderstanding that?**

**Comment by chughes297** [comment]
I would rather keep it separate, if possible. 

Two reasons: 

1) I think of inactive ballots as those that participated in the contest and then ran out of valid rankings for XYZ reason. Undervotes are votes that never went inactive, people skipped the contest entirely. 
2) This also reduces the number of ballots that will be counted as inactive lol.

**Comment by artoonie** [comment]
Sounds good. At least that verifies my understanding!

**Comment by artoonie** [comment]
A question about how to sort inactive ballots.

Take the following cast vote record:
1. Apple
2. Banana
3. Banana
4. Cherry

If Overvote Decision is set to `Exhaust`, this ballot becomes `inactive-by-overvote` once Apple is eliminated. If it`s set to `Skip To Next Rank`, it becomes `inactive-by-exhausted-choices` after Cherry is eliminated.

But what about if there is no fourth ranking for Cherry and Overvote Decision is `Skip To Next Rank`:
1. Apple
2. Banana
3. Banana

When banana is eliminated, does this ballot become `inactive-by-overvote` or `inactive-by-exhausted-choices`? I'm inclined to say `inactive-by-exhausted-choices` but I can see an argument for either.

**Comment by ggilber1** [comment]
It would be an inactive by exhausted choices. There is only an overvote if
the same ranking is assigned to two different candidates.

**Comment by artoonie** [comment]
Apologies, the original comment's example was incorrect -- it has been updated on github, which isn't reflected in the email notifications.

The updated example is here: https://github.com/BrightSpots/rcv/issues/633#issuecomment-1596197857

Specifically:
1: Apple
2: Banana
2: Cherry

**Comment by ggilber1** [comment]
Can't access the github right now, but I still think if the rule were
advance to next on an overvote, the ballot would become inactive due to
exhausted choices.

**Comment by artoonie** [comment]
Sounds good!

It makes sense to update the JSON to reflect these changes too. Here's my proposal. Changes marked with `####`.

```
{
  "config" : {
    "version", "1", #########
    "electionType", "Single-Winner", #########
    "contest" : "Aliases test (ES&S XLSX Format)",
    "date" : "2023-04-28",
    "jurisdiction" : "",
    "office" : ""
  },
  "summary": { #########
    "threshold" : 5, #########
    "numWinners": 1, #########
    "numCandidates": 4, #########
    "numVotesCast": 296077, #########
    "undervotes": 6453 #########
  }, #########
  "results" : [ {
    "round" : 1,
    "threshold": 5, #########
    "tally" : {
      "A" : "3",
      "B-2" : "3",
      "C" : "2",
      "D" : "1"
    },
    "tallyResults" : [ {
      "eliminated" : "D",
      "transfers" : {
        "exhausted" : "1"
      }
    } ]
  }, {
    "round" : 2,
    "threshold": 5, #########
    "tally" : {
      "A" : "3",
      "B-2" : "3",
      "C" : "2"
    },
    "tallyResults" : [ {
      "eliminated" : "C",
      "transfers" : {
        "A" : "2"
      }
    } ]
  }, {
    "round" : 3,
    "threshold": 5, #########
    "tally" : {
      "A" : "5",
      "B-2" : "3"
    },
    "tallyResults" : [ {
      "elected" : "A",
      "transfers" : { }
    } ]
  } ]
}
```

Any requests here?

Note: `version` is the config file version, not the tabulator version. This is version 1, and the previous version is implicitly version 0. This makes life easier for outside producers and consumers of this file format (of which there are several). The downside is that we'd now have an implicit mapping of Tabulator Version to Config Version.

It might also be reasonable to add a `generatedBy` field, where we add `RCTab Version 1.4.0`, to allow other producers to put their own field there (e.g. `RankedVote.co`)._

@chughes297 @HEdingfield would love input on this JSON format.

**Comment by tarheel** [comment]
> Note: version is the config file version, not the tabulator version. This is version 1, and the previous version is implicitly version 0. This makes life easier for outside producers and consumers of this file format (of which there are several). The downside is that we'd now have an implicit mapping of Tabulator Version to Config Version.

When you say "version is the config file version," do you mean that it's the JSON output version? Within the RCTab app we use the term "config" to mean the tabulation config files that the app takes as input.

Anyway, yeah, I agree with adding a `generatedBy` field also. And I'd call the output version something like `outputFormatVersion`, maybe?

**Comment by artoonie** [comment]
Yes, I meant JSON output version -- RCTab's outputs are RCVis' inputs, so the two get jumbled in my head :)

If we move `version` to the JSON root, I think leaving it as just `version` will make sense and avoid the need to rename to `outputFormatVersion`. What do you think? My reasoning is that for most users of the format (e.g. for RCVis, it'd be more like `inputVersion`; for RankedVote, it'd be more like `exportVersion`, and for RCVFormats, it'd be closer to `schemaVersion`). Comment above updated to reflect this.

**Comment by tarheel** [comment]
> Yes, I meant JSON output version -- RCTab's outputs are RCVis' inputs, so the two get jumbled in my head :)
> 
> If we move `version` to the JSON root, I think leaving it as just `version` will make sense and avoid the need to rename to `outputFormatVersion`. Comment above updated to reflect this. What do you think?
> 
> My reasoning is that for most users of the format, it's not an output: e.g. for RCVis, it'd be more like `inputVersion`; for RankedVote, it'd be more like `exportVersion`, and for RCVFormats, it'd be closer to `schemaVersion`.

Got it. I still think it's kind of confusing to call it simply `version`.  Maybe `jsonFormatVersion`? I just want something that clearly distinguishes it from the app version.

**Comment by artoonie** [comment]
That works for me!

**Comment by artoonie** [comment]
[Demo JSON](https://github.com/BrightSpots/rcv/blob/f18e73c6ee307891ad49499742883d0a99913f05/src/test/resources/network/brightspots/rcv/test_data/2013_minneapolis_mayor_scale/2013_minneapolis_mayor_scale_expected_summary.json)
[Demo CSV](https://docs.google.com/spreadsheets/d/10ZDn-CZK1f98lLZvct3ZqsuJwaLML29tvd3XAM_8nkk/edit#gid=591082084)

Can @chughes297 confirm these two files look as expected?

**Comment by chughes297** [comment]
Oh I see why they don't match up - the test file is using more aggressive overvote and skipped ranking rules than Minneapolis uses. That seems fine. 

Two remaining issues that stand out to me: 
1) I expected the "Total number of votes" field to report all ballots cast in the contest, not just those with a valid vote. 
2) There are too many undervotes. There were 80,101 total ballots cast in this contest (that's how many rows with vote info are in the CVR used for this test). There were 79,312 ballots with a valid ranking in this tabulation, and RCTab reports 4351 undervotes. That gives you 83,663 total ballots counted, which is incorrect. 
3) Inactive ballot totals don't make sense, either - RCTab reports 103 overvote exhuasts and 47 skipped ranking exhausts, but then says there are 789 inactive ballots - it should be 150.

**Comment by artoonie** [comment]
Some of these comments addressed on the Sheet comments:

1. Done
2. You're right -- I was overcounting undervotes.
3. Inactive ballots was including undervote counts -- fixed.
4. Continuing ballots was looking at "how many of this round continues to next;" renamed to "Active Ballots" as you suggested.
5. Changed "Total number of votes" to "Total number of ballots" and included inactive ballots in the count (and did not include undervotes).

Updated CSV is here: https://docs.google.com/spreadsheets/d/10ZDn-CZK1f98lLZvct3ZqsuJwaLML29tvd3XAM_8nkk/edit#gid=138274552

**Comment by artoonie** [comment]
@chughes297 the current codebase considers too-many-skipped-rankings as an undervote, and the GUI has a notion of an "Explicit Undervote Label" for that case.

We seem to be using Undervote in two different ways, one as part of this new feature (where Undervote = no rankings at all), and one in the current GUI (where Undervote = too many skipped rankings).

**Comment by chughes297** [comment]
The undervote label field in the GUI is for ES&S style CVRs, right? It allows users to input the term used to indicate a skipped ranking in an ES&S CVR.  We probably need to change that label (and the associated hint) to "Skipped ranking" or "Unused ranking" to avoid terminological confusion.

**Comment by yezr** [comment]
See #780 for a clarification on [this comment](https://github.com/BrightSpots/rcv/issues/633#issuecomment-1596197857) re: permissable ballot errors leading to an exhausted ballot.
